torchvision/csrc/io/image/cpu/decode_gif.cpp

```diff
@@ -0,0 +1,136 @@
+#include "decode_gif.h"
+#include "giflib/gif_lib.h"
+
+namespace vision {
+namespace image {
+
+typedef struct reader_helper_t {
+  torch::Tensor const* encoded_bytes;
+  int num_bytes_read; // number of bytes read so far in the tensor
+} reader_helper_t;
+
+// That function is used by GIFLIB routines to read the encoded bytes.
+// This reads `len` bytes and writes them into `buf`. The data is read from the
+// input tensor passed to decode_gif().
+int read_from_tensor(GifFileType* gifFile, GifByteType* buf, int len) {
+  // the UserData field was set in DGifOpen()
+  reader_helper_t* reader_helper = (reader_helper_t*)gifFile->UserData;
+  auto num_encoded_bytes = (int)reader_helper->encoded_bytes->numel();
+  auto data_ptr = reader_helper->encoded_bytes->data_ptr<GifByteType>();
+
+  auto i = 0;
+  auto num_bytes_to_read =
+      std::min(len, num_encoded_bytes - reader_helper->num_bytes_read);
+  while (i < num_bytes_to_read) {
+    buf[i] = data_ptr[reader_helper->num_bytes_read + i];
+    i++;
+  }
+  reader_helper->num_bytes_read += i;
+  return i;
+}
+
+torch::Tensor decode_gif(const torch::Tensor& encoded_bytes) {
+  // LibGif docs: https://giflib.sourceforge.net/intro.html
+  // Refer over there for more details on the libgif API, API ref, and a
+  // detailed description of the GIF format.
+
+  int error = D_GIF_SUCCEEDED;
+
+  // We're using DGidOpen. The other entrypoints of libgif are
+  // DGifOpenFileName and DGifOpenFileHandle but we don't want to use those,
+  // since we need to read the encoded bytes from a tensor of encoded bytes, not
+  // from a file (for consistency with existing jpeg and png decoders). Using
+  // DGifOpen is the only way to read from a custom source.
+  // For that we need to provide a reader function `read_from_tensor` that
+  // reads from the tensor, and we have to keep track of the number of bytes
+  // read so far: this is why we need the reader_helper struct.
+
+  // TODO: We are potentially doing an unnecessary copy of the encoded bytes:
+  // - 1 copy in from file to tensor (in read_file())
+  // - 1 copy from tensor to GIFLIB buffers (in read_from_tensor())
+  // Since we're vendoring GIFLIB we can potentially modify the calls to
+  // InternalRead() and just set the `buf` pointer to the tensor data directly.
+  // That might even save allocation of those buffers.
+  // If we do that, we'd have to make sure the buffers are never written to by
+  // GIFLIB, otherwise we'd be overridding the tensor data.
```

Comment:
Leaving this out for now, we can try to address it if gif decoding ever becomes a bottleneck...